File: src/visual-graph/builder/loop-test-node-id.test.ts
import { describe, expect, test } from "vitest";

import {
  doWhileTestNodeId,
  forTestNodeId,
  whileTestNodeId,
} from "./loop-test-node-id.js";

describe("whileTestNodeId", () => {
  test("composes 'while_test_<scope>_<offset>'", () => {
    expect(whileTestNodeId("scope_0", 33)).toBe("while_test_scope_0_33");
  });

  test("sanitises non-alphanumeric characters in the parent scope id", () => {
    expect(whileTestNodeId("scope.0:nested", 7)).toBe(
      "while_test_scope_0_nested_7",
    );
  });

  test("accepts an empty parent scope id", () => {
    expect(whileTestNodeId("", 12)).toBe("while_test__12");
  });
});

describe("doWhileTestNodeId", () => {
  test("composes 'do_while_test_<scope>_<offset>'", () => {
    expect(doWhileTestNodeId("scope_0", 41)).toBe("do_while_test_scope_0_41");
  });

  test("sanitises non-alphanumeric characters in the parent scope id", () => {
    expect(doWhileTestNodeId("a-b", 5)).toBe("do_while_test_a_b_5");
  });
});

describe("forTestNodeId", () => {
  test("composes 'for_test_<scope>_<offset>'", () => {
    expect(forTestNodeId("scope_1", 99)).toBe("for_test_scope_1_99");
  });

  test("sanitises non-alphanumeric characters in the parent scope id", () => {
    expect(forTestNodeId("scope#2!", 0)).toBe("for_test_scope_2__0");
  });
});

File: crates/unsnarl-visual-graph/src/builder/loop_test_node_id_test.rs
//! Sibling tests for [`while_test_node_id`], [`do_while_test_node_id`],
//! and [`for_test_node_id`].

use super::{do_while_test_node_id, for_test_node_id, while_test_node_id};

#[test]
fn while_test_composes_prefix_scope_offset() {
    assert_eq!(while_test_node_id("scope_0", 33), "while_test_scope_0_33");
}

#[test]
fn while_test_sanitises_non_alphanumerics_in_scope_id() {
    assert_eq!(
        while_test_node_id("scope.0:nested", 7),
        "while_test_scope_0_nested_7"
    );
}

#[test]
fn while_test_accepts_empty_parent_scope_id() {
    assert_eq!(while_test_node_id("", 12), "while_test__12");
}

#[test]
fn do_while_test_composes_prefix_scope_offset() {
    assert_eq!(
        do_while_test_node_id("scope_0", 41),
        "do_while_test_scope_0_41"
    );
}

#[test]
fn do_while_test_sanitises_non_alphanumerics_in_scope_id() {
    assert_eq!(do_while_test_node_id("a-b", 5), "do_while_test_a_b_5");
}

#[test]
fn for_test_composes_prefix_scope_offset() {
    assert_eq!(for_test_node_id("scope_1", 99), "for_test_scope_1_99");
}

#[test]
fn for_test_sanitises_non_alphanumerics_in_scope_id() {
    assert_eq!(for_test_node_id("scope#2!", 0), "for_test_scope_2__0");
}
